cmd/id-exporter/main.go

```diff
@@ -48,18 +86,54 @@ func (c idExporter) findIDs() ([]id, error) {
 		c.log.AuditErrf("Error finding IDs: %s", err)
 		return nil, err
 	}
+	return holder, nil
+}
 
-	return idsList, nil
+// Find all registration IDs with unexpired certificates and gather an
+// example hostname.
+func (c idExporter) findIDsWithExampleHostnames() (idExporterResults, error) {
+	var holder idExporterResults
+	_, err := c.dbMap.Select(
+		// A previous version of this query resulted in an off-by-one
+		// error, the resulting change removed `ANY_VALUE(...)` for the
+		// hostname. This query now relies upon undefined behavior.
+		// Though, the use of `ANY_VALUE` was itself a defined but
+		// nondeterministic behavior.
```

Comment:
This comment is confusing because the "previous version" is not visible here. In general when putting a comment in place for the long haul, rather than phrase things in terms of what used to be, it's better to phrase them in terms of "here's why we can't do it X way." In other words, think of the mental state of your future reader: They're not thinking about what the previous query looked like. They're thinking about how else they might implement it.

Also "undefined behavior" is pretty scary! Can you provide an explanation of why it's safe enough to use here?

Reply:
While reasoning through this I realized that I only yanked this from the writeup that @jcjones did because I wanted a future engineer to avoid introducing the off-by-one error present in the original query. However, we've got tests for that. We can just get rid of this.

@jcjones should probably be the one to address why this query is safe / acceptable though.

Reply:
I think the right way to think about this is that it is undefined and forbidden behavior per the SQL specification, but is a [deliberate nonstandard behavior in MySQL/MariaDB](https://dev.mysql.com/doc/refman/5.7/en/group-by-handling.html). 

This behavior is being phased out; MySQL prohibited it by default in the [same release that they added the SQL-standard ANY_VALUE aggregate function](https://dev.mysql.com/doc/relnotes/mysql/5.7/en/news-5-7-5.html#mysqld-5-7-5-sql-mode). 

Since ANY_VALUE has been available — and SELECTING a nonaggregated value prohibited — in MySQL for over six years, I had assumed the same was true for MariaDB. But MariaDB doesn't prohibit it, instead continuing the earlier MySQL behavior.

Perhaps a better comment might be like:
> This query relies upon nonstandard SQL behavior as reversedName is not aggregated via function or via the GROUP BY. In proper SQL this would be done using the ANY_VALUE function, but until that function is available, this function will rely on the SQL Mode [ONLY_FULL_GROUP_BY](https://mariadb.com/kb/en/sql-mode/#only_full_group_by) being left disabled.

Reply:
Great, thanks! I am satisfied. This looks good, pending resolution of your nit below.